caluma/caluma_form/jexl.py

```diff
@@ -103,93 +94,96 @@ def use_question_context(self, question_slug):
 
-        If all subforms are hidden where the question shows up,
-        the question shall be hidden as well.
-        """
-        paths = self._structure.root().paths_to_question(question.slug)
-
-        res = bool(paths) and all(
-            # the "inner" check here represents a single path. If any
-            # question is hidden, the question is not reachable via
-            # this path
-            bool(path) and any(self.is_hidden(fq) for fq in path if fq != question)
-            for path in paths
-        )
-        # The outer check verifies if all paths are "hidden" (ie have a hidden
-        # question in them). If all paths from the root form to the
-        # question are hidden, the question must indeed be hidden itself.
-
-        return res
+        for slug in dependencies:
+            if slug not in referenced_slugs:
+                raise QuestionMissing(
+                    f"Question `{slug}` could not be found in form {field.form}"
+                )
 
-    def _cache_key(self, question_slug):
-        field = self._structure.get_field(question_slug)
-        return (field.document.pk, question_slug)
+        return referenced_fields
 
-    def is_hidden(self, question):
-        """Return True if the given question is hidden.
+    def is_hidden(self, field: Field):
+        """Return True if the given field is hidden.
 
         This checks whether the dependency questions are hidden, then
-        evaluates the question's is_hidden expression itself.
+        evaluates the field's is_hidden expression itself.
         """
-        cache_key = self._cache_key(question.pk)
+        cache_key = (field.document.pk, field.question.pk)
 
         if cache_key in self._cache["hidden"]:
             return self._cache["hidden"][cache_key]
+
         # Check visibility of dependencies before actually evaluating the `is_hidden`
         # expression. If all dependencies are hidden,
         # there is no way to evaluate our own visibility, so we default to
         # hidden state as well.
-
-        deps = list(self.extract_referenced_questions(question.is_hidden))
+        deps = list(self.extract_referenced_questions(field.question.is_hidden))
+        referenced_fields = self._get_referenced_fields(field, deps)
 
         # all() returns True for the empty set, thus we need to
         # check that we have some deps at all first
         all_deps_hidden = bool(deps) and all(
-            self.is_hidden(self._question(dep)) for dep in deps
+            self.is_hidden(ref_field) for ref_field in referenced_fields
         )
         if all_deps_hidden:
             self._cache["hidden"][cache_key] = True
             return True
 
         # Also check if the question is hidden indirectly,
         # for example via parent formquestion.
-        if self._all_containers_hidden(question):
+        parent = field.parent()
+        if parent and parent.question and self.is_hidden(parent):
```

Comment:
Why is `all_containers_hidden` not needed anymore? Do we assume that there is always only one path to a given question?

Reply:
Because we handle this in `is_hidden()` via recursion now. This is possible because we always look from fields, not from questions anymore. `is_required()` did this partially already.

Reply:
Alright, thanks!